File: src/main/java/cn/keevol/keenotes/mobilefx/PosterInkTheme.java
package cn.keevol.keenotes.mobilefx;

/**
 * Ink wash themes for note posters. Mirrors the iOS poster theme choices.
 */
public enum PosterInkTheme {
    MEI("PosterInkMei", "梅", "/poster/ink/PosterInkMei.png"),
    LAN("PosterInkLan", "兰", "/poster/ink/PosterInkLan.png"),
    ZHU("PosterInkZhu", "竹", "/poster/ink/PosterInkZhu.png"),
    JU("PosterInkJu", "菊", "/poster/ink/PosterInkJu.png");

    private static final PosterInkTheme[] VALUES = values();

    private final String assetName;
    private final String label;
    private final String resourcePath;

    PosterInkTheme(String assetName, String label, String resourcePath) {
        this.assetName = assetName;
        this.label = label;
        this.resourcePath = resourcePath;
    }

    public String getAssetName() {
        return assetName;
    }

    public String getLabel() {
        return label;
    }

    public String getResourcePath() {
        return resourcePath;
    }

    public static PosterInkTheme stableForNoteId(long noteId) {
        int index = (int) Math.floorMod(noteId, VALUES.length);
        return VALUES[index];
    }

    public PosterInkTheme next() {
        int nextIndex = (ordinal() + 1) % VALUES.length;
        return VALUES[nextIndex];
    }

    public float opacityFor(int contentLength) {
        if (contentLength > 900) {
            return 0.07f;
        }
        if (contentLength > 500) {
            return 0.09f;
        }
        if (contentLength > 220) {
            return 0.10f;
        }
        return 0.11f;
    }
}

File: keenotes-android/app/src/main/java/cn/keevol/keenotes/share/PosterInkTheme.kt
package cn.keevol.keenotes.share

import androidx.annotation.DrawableRes
import cn.keevol.keenotes.R
import kotlin.math.roundToInt

enum class PosterInkTheme(
    @DrawableRes val drawableRes: Int,
    val label: String
) {
    MEI(R.drawable.poster_ink_mei, "梅"),
    LAN(R.drawable.poster_ink_lan, "兰"),
    ZHU(R.drawable.poster_ink_zhu, "竹"),
    JU(R.drawable.poster_ink_ju, "菊");

    fun next(): PosterInkTheme {
        val values = entries.toTypedArray()
        return values[(ordinal + 1) % values.size]
    }

    fun alphaFor(contentLength: Int): Int {
        val opacity = when {
            contentLength > 900 -> 0.07f
            contentLength > 500 -> 0.09f
            contentLength > 220 -> 0.10f
            else -> 0.11f
        }
        return (opacity * 255).roundToInt().coerceIn(0, 255)
    }

    companion object {
        fun stableForNoteId(noteId: Long): PosterInkTheme {
            val values = entries.toTypedArray()
            val index = Math.floorMod(noteId, values.size.toLong()).toInt()
            return values[index]
        }
    }
}
